### PR TITLE
[gen-l10n] NNBD generated code

### LINKED ISSUE #68598: [gen-l10n] Make generated code null safe
Once the `intl` package is migrated to NNBD and the Flutter framework is completely migrated to be null safe, we will likely need to create a new option to specify whether or not to generate null-safe code for `flutter gen-l10n`. This is particularly important for Flutter apps that plan to use NNBD since the `gen-l10n` tool currently isn't null safe.

Proposal: Update gen-l10n to use NNBD, but create an option to generate non-null safe code for backwards compatibility as some users may require more time to carry over.

cc/ @HansMuller @pennzht

### PR DESCRIPTION
**Pull request by shihaohong**
## Description

Migrates the `flutter gen-l10n` generated code to be NNBD.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/68598

## Tests

I added the following tests:

- Updated the integration test to use NNBD

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart [hunk 1]
```diff
@@ -35,7 +35,7 @@ class GenL10nProject extends Project {
   final String pubspec = '''
 name: test
 environment:
-  sdk: ">=2.0.0-dev.68.0 <3.0.0"
+  sdk: ">=2.11.0-213.0.dev <2.12.0"
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart:38]: 2.12.0-0

**Review by jonahwilliams — COMMENTED** [review]
You also need to change the value here to 2.12:

https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/dart/pub.dart#L458

File: packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart:38]: (i.e. this should run in null safe mode so we're sure that it works)

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Comment by shihaohong** [comment]
The failing Google testing tests do not seem related to my change, but they do seem related to NNBD. I'll hold off on merging until it seems safe to do so.

**Pull request merged**